### DISCUSSION TITLE
RAM compression -  1GB of ZRAM amounts to around 2 to 3 GB after compression.

### DISCUSSION THREAD

**Discussion by kcramakrishna** [Digital-Assistant/Digital_Assistant_Client#310] [General]
zRAM and zCache are really good ways to more efficiently use RAM. 

Apparently you get between 2.5x to 3.9x "extra" RAM. Here are some numbers:
https://old.reddit.com/r/linuxquestions/comments/ju4bft/zram_zswap_or_both/gccnqby/

zRAM vs. zCache vs. zSwap: https://www.maketecheasier.com/zram-zcache-zswap/ 
(Aparently zSwap is better but zRAM is simpler)

zSWAP vs. zRAM:
https://stackoverflow.com/questions/18437205/difference-between-zram-and-zswap
https://askubuntu.com/questions/471912/zram-vs-zswap-vs-zcache-ultimate-guide-when-to-use-which-one
https://newbedev.com/zram-vs-zswap-vs-zcache-ultimate-guide-when-to-use-which-one
https://old.reddit.com/r/linuxquestions/comments/ju4bft/zram_zswap_or_both/

zSWAP is sorta experimental: (https://www.kernel.org/doc/html/v4.18/vm/zswap.html)
zRAM is more mature but has lesser features: https://www.kernel.org/doc/html/latest/admin-guide/blockdev/zram.html

zSWAP is especially useful when you have dockers/Kubes etc. I think.
We should use zSWAP if you have SWAP on your system. If you do not have SWAP, then use zRAM.

Some zRAM articles:
(zRAM on Ubuntu): https://www.maketecheasier.com/configure-zram-ubuntu/ 
https://www.cnx-software.com/2018/05/14/running-out-of-ram-in-ubuntu-enable-zram/
https://www.techrepublic.com/article/how-to-enable-the-zram-module-for-faster-swapping-on-linux/
https://haydenjames.io/linux-performance-almost-always-add-swap-part2-zram/

increasing SWAP in Ubuntu: 
https://bogdancornianu.com/change-swap-size-in-ubuntu/

On a 16GB machine, dedicating 10GB of RAM to zram can effectively give you (10 x 2.5) + 6 ~ +30GB of effective RAM. This is useful for most development machines.

**Comment by kcramakrishna** [comment]
Configuration options:
1. zSWAP: 
https://www.maketecheasier.com/use-zswap-improve-old-linux-performance/
https://ubuntu-mate.community/t/enable-zswap-to-increase-performance/11302
1.a Make sure you have SWAP equal to your RAM (general best practice). i.e. configure 16GB SWAP if you have 16GB RAM. Here is how you can increase/decrease SWAP: https://bogdancornianu.com/change-swap-size-in-ubuntu/  (Historically we used to have 2 times SWAP as RAM. With modern systems with high RAM, this came down to 1:1 of SWAP:RAM. With high end systems with a LOT of RAM, some do not have SWAP at all)
1.b Enable zSWAP https://www.kernel.org/doc/html/v4.18/vm/zswap.html (Check on how to do this in your distro)
1.c You should be fine with default settings for compressions etc.

2. zRAM: (If you do not have SWAP or if you don't want SWAP. zRAM is also easier to setup and configure.)
2.a: Enable zRAM as per your distro. For Ubuntu - just add the package and reboot. Apparently is configures itself with default params which should be OK in most cases. (https://askubuntu.com/questions/174579/how-do-i-use-zram)
To activate it, type in terminal:

sudo apt-get install zram-config

To remove [type the following]:

sudo dpkg --purge zramswap-enabler
sudo dpkg --purge zram-config

2.b: You can tweak the zRAM params if you are more adventurous and want to eke out more performance (i.e. more RAM dedicated to zRAM) - "zramctl --help" (https://www.cnx-software.com/2018/05/14/running-out-of-ram-in-ubuntu-enable-zram/)
You can do a lot of tweaking depending on your workload and system (processor/RAM) - for newer (or powerful) processors, you can configure high compression algorithms etc.
https://www.maketecheasier.com/configure-zram-ubuntu/
2.c zRAM has one disadvantage - it does not swap out stale pages. If you shutdown your laptop at the end of the day, you will be fine with zRAM. If you reboot very infrequently (eg. servers), then you might want to chose zSWAP over zRAM.

Congratulations on having doubled your RAM capacity without having to buy physical RAM.

**Comment by yureshwar** [comment]
I have enabled the zram and my swap size got increased from 16gb to 25.4gb

Also showing the zram swaps and algorithm which is being used.

![image](https://user-images.githubusercontent.com/516516/146411412-e5316a93-03e3-41eb-992b-da97129809f0.png)

![image](https://user-images.githubusercontent.com/516516/146411613-9bf6446f-de7b-434d-a2d1-2adcf52a48c1.png)

**Comment by kcramakrishna** [comment]
https://ubuntu.com/blog/how-low-can-you-go-running-ubuntu-desktop-on-a-2gb-raspberry-pi-4

**Comment by yureshwar** [comment]
![image](https://user-images.githubusercontent.com/516516/164286149-bfa24e4e-e59d-4e89-a5a0-f4c120d642ae.png)
After enabling zswap

**Comment by kcramakrishna** [comment]
We will use zram based on this information:
https://g.co/bard/share/bc32847872ee 

Summary:
1. Maturity: zram is more mature.
2. Efficiency: When we run out of RAM, zram writes compressed pages to SWAP. compressed pages will be more efficient due to smaller size for io and processor savings since we do not have to uncompress the pages. zswap on the other hand decompresses pages before writing to SWAP.


We will also enable swapspace daemon: 
https://pqxx.org/swapspace/
https://manpages.ubuntu.com/manpages/trusty/man8/swapspace.8.html

Summary:
Swapspace dynamically increases/decreases SWAP size on disk based on demand. This helps avoid having to size SWAP during OS installation.

**Comment by yureshwar** [comment]
This is a good doc on what we can do on zram.

https://docs.kernel.org/admin-guide/blockdev/zram.html
